### PR TITLE
SLP/RISCV: add test for vectorized ctpop, like in X86

### PR DESCRIPTION
**Pull request by artagnon**
Recently, 7f26c27 turned on SLP by default for RISC-V, and although
there are quite a few tests for SLP under the X86/ target, it is unclear
whether the same constructs would be vectorized on RISC-V. This patch
takes a step in the direction of remedying this, by noticing that ctpop
is often vectorized on RISC-V, and adding four tests for different
integer widths.

### CODE REVIEW

**Comment by alexey-bataev** [comment]
Would be good to add the corresponding test for the analysis/cost model

**Comment by preames** [comment]
Please rework your commit message.  If your goal is to exercise the costs, then you really should be writing a cost model test, not an SLP test.  I suspect that despite your wording, your actual intention is to add coverage showing that SLP can vectorize these examples on RISCV.  (The difference in wording there is that the cost of the ctpop expansion is only part of that.)

Note that we do appear to have cost model tests already in test/Analysis/CostModel/RISCV/int-bit-manip.ll.

**Comment by artagnon** [comment]
Thanks @preames; re-worded.

File: llvm/test/Transforms/SLPVectorizer/RISCV/ctpop.ll [hunk 1]
```diff
@@ -0,0 +1,107 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 2
+; RUN: opt < %s -passes=slp-vectorizer -S -mtriple=riscv64 -mattr=+v,+f | FileCheck %s
```
Comment (RKSimon) [inline llvm/test/Transforms/SLPVectorizer/RISCV/ctpop.ll:2]: Test more attribute variants?

Comment (RKSimon) [reply llvm/test/Transforms/SLPVectorizer/RISCV/ctpop.ll:2]: I don't know much about riscv, but maybe similar RUNs as found in the rvv/fixed-vectors-ctpop.ll codegen tests?

**Comment by eopXD** [comment]
Why specifically `ctpop`. There are much more coverage for the X86 backend, is the whole plan to add them commit by commit to the RISC-V backend?

**Comment by artagnon** [comment]
> Why specifically `ctpop`.

I was investigating some benchmarks where I came across `ctpop` being vectorized by gcc-aarch64, but not llvm-riscv, so I was curious about whether LLVM does any vectorization of `ctpop` at all on RISC-V. This patch is a result of that investigation; nothing more.

> There are much more coverage for the X86 backend, is the whole plan to add them commit by commit to the RISC-V backend?

Do you have a different suggestion?

File: llvm/test/Transforms/SLPVectorizer/RISCV/ctpop.ll (hunk 1 above)
Comment (artagnon) [reply llvm/test/Transforms/SLPVectorizer/RISCV/ctpop.ll]: I added a bunch of RUN lines, although the CHECK prefixes are unused: the codegen is identical. I hope this resolves the issue.

File: llvm/test/Transforms/SLPVectorizer/RISCV/ctpop.ll [hunk 2]
```diff
@@ -0,0 +1,115 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 2
+; RUN: opt < %s -passes=slp-vectorizer -S -mtriple=riscv32 -mattr=+m,+v | FileCheck %s --check-prefixes=CHECK,RV32
+; RUN: opt < %s -passes=slp-vectorizer -S -mtriple=riscv64 -mattr=+m,+v | FileCheck %s --check-prefixes=CHECK,RV64
+; RUN: opt < %s -passes=slp-vectorizer -S -mtriple=riscv32 -mattr=+v,+experimental-zvbb | FileCheck %s --check-prefixes=CHECK,ZVBB-RV32
+; RUN: opt < %s -passes=slp-vectorizer -S -mtriple=riscv64 -mattr=+v,+experimental-zvbb | FileCheck %s --check-prefixes=CHECK,ZVBB-RV64
```
Comment (alexey-bataev) [inline llvm/test/Transforms/SLPVectorizer/RISCV/ctpop.ll:5]: Drop unused check labels and remove extra comments at the end of the file, generated by the script

**Review by alexey-bataev — APPROVED** [review]

**Comment by preames** [comment]
LGTM as well.

For your follow up on this, note that vector ctpop exists only under ZVBB.  I suspect we didn't update the costs for ctpop if zvbb is enabled.  Without it, vectorizing is probably not profitable with the default expansion.  That'd be the first thing I'd check.  If my guess is right, you're also want a CostModel test.  

@lukel97 FYI

**Pull request merged**